### PR TITLE
Address #54 and improve interrupted shutdown

### PR DESCRIPTION
**Pull request by romain-intel**
Although not fully guaranteeing a clean shutdown every time, this patch makes the shutdown a little more robust (giving time to children process to shutdown), prevents multiple CTRL-C from interrupting the shutdown (in most cases) and is possibly faster in the presence of large fan outs.

### CODE REVIEW

File: metaflow/runtime.py [hunk 1]
```diff
@@ -216,6 +216,13 @@ def execute(self):
 
         except KeyboardInterrupt as ex:
             self._logger('Workflow interrupted.', system_msg=True, bad=True)
+            # Ignore new signals
+            import signal
+
+            def print_ignore(signum, frame):
+                self._logger('Shutdown in progress -- ignoring additional CTRL-C',
```
Comment (savingoyal) [inline metaflow/runtime.py]: It will be nice if the second `ctrl-C` instantly terminates the execution instead of blocking on the shutdown procedure.

File: metaflow/runtime.py [hunk 2]
```diff
@@ -238,8 +245,32 @@ def execute(self):
...
+        # If we are here, all children have received a signal and are shutting down.
+        # We want to give them an opportunity to do so and then kill
+        live_workers = []
+        temp_workers = {}
         for worker in self._workers.values():
-            worker.kill()
+            temp_workers[worker] = 1
+        live_workers = temp_workers.keys()
+        now = int(time.time())
+        self._logger('Waiting for %d workers to shutdown...' % len(live_workers),
```
Comment (savingoyal) [inline metaflow/runtime.py]: The notion of workers is not exposed to users of Metaflow. Maybe reword this or take this out entirely?

File: metaflow/runtime.py [hunk 3]
```diff
@@ -826,16 +858,22 @@ def fds(self):
         return (self._proc.stderr.fileno(),
                 self._proc.stdout.fileno())
 
-    def kill(self):
-        if not self.killed:
+    def pre_kill(self):
```
Comment (savingoyal) [inline metaflow/runtime.py]: maybe `clean` instead of `pre_kill` and `self.cleaned` instead of `self.prekilled`?

File: metaflow/runtime.py [hunk 4]
```diff
@@ -238,8 +245,32 @@ def execute(self):
...
 
     def _killall(self):
+        # If we are here, all children have received a signal and are shutting down.
+        # We want to give them an opportunity to do so and then kill
+        live_workers = []
+        temp_workers = {}
         for worker in self._workers.values():
-            worker.kill()
+            temp_workers[worker] = 1
+        live_workers = temp_workers.keys()
```
Comment (savingoyal) [inline metaflow/runtime.py]: isn't `live_workers = self._workers.values()`?

File: metaflow/runtime.py [hunk 5]
```diff
@@ -238,8 +245,32 @@ def execute(self):
...
+        now = int(time.time())
+        self._logger('Waiting for %d workers to shutdown...' % len(live_workers),
+                     system_msg=True, bad=True)
+        while live_workers and int(time.time()) - now < 5:
+            # While not all workers are dead and we have waited less than 5 seconds
+            new_live_workers = []
+            for worker in live_workers:
+                if not worker.pre_kill():
+                    new_live_workers.append(worker)
+            live_workers = new_live_workers
```
Comment (savingoyal) [inline metaflow/runtime.py]: minor syntactic nit (up to you to incorporate it) - `live_workers = [worker for worker in live_workers if not worker.pre_kill()]`

File: metaflow/runtime.py (hunk 2 above)
Comment (savingoyal) [reply metaflow/runtime.py]: Similar for L:266 and L:272.

File: metaflow/runtime.py (hunk 5 above)
Comment (savingoyal) [reply metaflow/runtime.py]: Also, is the intended behavior that every worker gets the `pre_kill` signal at least once, even if it takes more than 5 seconds?

File: metaflow/runtime.py (hunk 1 above)
Comment (romain-intel) [reply metaflow/runtime.py]: I think before that's what it did. I wanted to try to prevent multiple CTRL-C from preventing us from shutting down. I can remove this though if you feel that behavior is better.

File: metaflow/runtime.py (hunk 4 above)
Comment (romain-intel) [reply metaflow/runtime.py]: Not quite; in workers, they appear twice (one for each FD). This dedups them.

File: metaflow/runtime.py (hunk 2 above)
Comment (romain-intel) [reply metaflow/runtime.py]: I can maybe say something like "Terminating %d running steps...". I do want to give some feedback to make sure the user feels something is happening.

File: metaflow/runtime.py (hunk 5 above)
Comment (romain-intel) [reply metaflow/runtime.py]: Yes, I thought of checking every time in the loop but decided against it because it would be less clean and nothing in the loop is blocking (they are all polls) so it should be relatively fast in all cases so it may take slightly more than 5 seconds but I doubt it will go longer.

File: metaflow/runtime.py (hunk 1 above)
Comment (savingoyal) [reply metaflow/runtime.py]: Yeah, let's remove it.

File: metaflow/runtime.py (hunk 4 above)
Comment (savingoyal) [reply metaflow/runtime.py]: so, `live_workers = set(self._workers.values())`

File: metaflow/runtime.py [hunk 6]
```diff
@@ -238,8 +245,28 @@ def execute(self):
...
+            # While not all workers are dead and we have waited less than 5 seconds
+            live_workers = [worker for worker in live_workers if not worker.clean()]
+        if live_workers:
+            self._logger('Killing remaining %d tasks after having waited for %d seconds -- '
+                         'some tasks may not exit cleanly' % (len(live_workers),
+                                                              int(time.time()) - now),
+                         system_msg=True, bad=True)
+            for worker in live_workers:
+                worker.kill()
+        self._logger('Flushing logs -- you may see messages from the tasks that were shutdown...',
```
Comment (savingoyal) [inline metaflow/runtime.py]: Should we remove this logging message?

File: metaflow/runtime.py [hunk 7]
```diff
@@ -238,8 +245,28 @@ def execute(self):
...
+            temp_workers[worker] = 1
+        live_workers = temp_workers.keys()
+        now = int(time.time())
+        self._logger('Terminating %d active tasks...' % len(live_workers),
+                     system_msg=True, bad=True)
+        while live_workers and int(time.time()) - now < 5:
+            # While not all workers are dead and we have waited less than 5 seconds
+            live_workers = [worker for worker in live_workers if not worker.clean()]
+        if live_workers:
+            self._logger('Killing remaining %d tasks after having waited for %d seconds -- '
```
Comment (savingoyal) [inline metaflow/runtime.py]: For consistency with L:256 - maybe just - `Killing %d remaining tasks...`

File: metaflow/runtime.py (hunk 6 above)
Comment (romain-intel) [reply metaflow/runtime.py]: I wanted to keep it to help users understand why they see printout after things have been waited for and terminated.

Comment (savingoyal) [reply metaflow/runtime.py]: I am worried that this log is too verbose. Maybe just? - `Flushing logs...`

File: metaflow/runtime.py (hunk 4 above)
Comment (seeravikiran) [reply metaflow/runtime.py]: Could you also detail what `temp_workers` is for? If it's for the sole purpose of creating the set, I'd instead do what @savingoyal suggests w.r.t set.

File: metaflow/runtime.py [hunk 8]
```diff
@@ -238,8 +238,28 @@ def execute(self):
...
         for worker in self._workers.values():
-            worker.kill()
+            temp_workers[worker] = 1
+        live_workers = temp_workers.keys()
+        now = int(time.time())
+        self._logger('Terminating %d active tasks...' % len(live_workers),
+                     system_msg=True, bad=True)
+        while live_workers and int(time.time()) - now < 5:
+            # While not all workers are dead and we have waited less than 5 seconds
+            live_workers = [worker for worker in live_workers if not worker.clean()]
```
Comment (seeravikiran) [inline metaflow/runtime.py]: Feel free to reject this suggestion - 

This seems like a good use-case for `filter`:
live_workers = filter(lambda worker: not worker.clean(), live_workers)
I don't know the mutability guarantees for list iteration, so I'm not sure if using a function like `filter` is better.

As bonus it might work on set(s) or list(s).

File: metaflow/runtime.py [hunk 9]
```diff
@@ -238,8 +238,23 @@ def execute(self):
                                         'by the end of flow.')
 
     def _killall(self):
-        for worker in self._workers.values():
-            worker.kill()
+        # If we are here, all children have received a signal and are shutting down.
+        # We want to give them an opportunity to do so and then kill
+        live_workers = list(set(self._workers.values()))
```
Comment (seeravikiran) [inline metaflow/runtime.py]: Why do you need this be a list? It seems you just need iteration, and the below suggested `filter`.

File: metaflow/runtime.py [hunk 10]
```diff
@@ -770,6 +785,7 @@ def __init__(self, task, max_logs_size):
 
         self._encoding = sys.stdout.encoding or 'UTF-8'
         self.killed = False
+        self.cleaned = False
```
Comment (seeravikiran) [inline metaflow/runtime.py]: Could you add a short comment on what self.killed=True represents vs self.cleaned=True?
It also seems if self.killed = True, then self.cleaned must be True?

Defining the state transition perhaps in a short comment helps understand why you are tracking two states.

File: metaflow/runtime.py (hunk 8 above)
Comment (romain-intel) [reply metaflow/runtime.py]: I read a bit that filter may be a tad slower (not a huge deal here though) but I find list comprehension to be more readable so if OK, I would prefer to leave it as such.

File: metaflow/runtime.py (hunk 9 above)
Comment (romain-intel) [reply metaflow/runtime.py]: sure, I can change it. It made more sense for it to be a list when I had a loop with add but now I can put just a set.
